libs/filer/workspace_files_extensions_client.go

```diff
@@ -0,0 +1,328 @@
+					commonName: path.Join(name, entries[i].Name()),
+				}
+			}
+			seenPaths[entry.Name()] = sysInfo
+			continue
+		}
+
+		// Add extension to local file path if the file is a notebook
+		stat, err := w.addNotebookExtension(ctx, entry.Name())
+		if err != nil {
+			return nil, err
+		}
+		entries[i] = wsfsDirEntry{wsfsFileInfo{oi: *stat.ObjectInfo}}
+
+		// Error if we have seen this path before in the current directory.
+		// If not seen before, add it to the seen paths.
+		if _, ok := seenPaths[entries[i].Name()]; ok {
+			return nil, DuplicatePathError{
+				oi1:        seenPaths[entries[i].Name()],
+				oi2:        sysInfo,
+				commonName: path.Join(name, entries[i].Name()),
+			}
+		}
+		seenPaths[entries[i].Name()] = sysInfo
+
+	}
+
+	return entries, nil
+}
+
+// Note: The import API returns opaque internal errors for namespace clashes
+// (e.g. a file and a notebook or a directory and a notebook). Thus users of this
+// method should be careful to avoid such clashes.
+func (w *workspaceFilesExtensionsClient) Write(ctx context.Context, name string, reader io.Reader, mode ...WriteMode) error {
+	return w.filer.Write(ctx, name, reader, mode...)
+}
+
+// Try to read the file as a regular file. If the file is not found, try to read it as a notebook.
+func (w *workspaceFilesExtensionsClient) Read(ctx context.Context, name string) (io.ReadCloser, error) {
+	r, err := w.filer.Read(ctx, name)
+
+	// If the file is not found, it might be a notebook.
+	if errors.As(err, &FileDoesNotExistError{}) {
+		stat, ok := w.removeNotebookExtension(ctx, name)
+		if !ok {
+			// Not a valid notebook. Return the original error.
+			return nil, err
+		}
+		// The workspace files filer performs an additional stat call to make sure
+		// the path is not a directory. We can skip this step since we already have
+		// the stat object and know that the path is a notebook.
+		return w.workspaceClient.Workspace.Download(
+			ctx,
+			path.Join(w.root, stat.nameForWorkspaceAPI),
+			workspace.DownloadFormat(stat.ReposExportFormat),
+		)
+	}
+	return r, err
+}
+
+// Try to delete the file as a regular file. If the file is not found, try to delete it as a notebook.
+func (w *workspaceFilesExtensionsClient) Delete(ctx context.Context, name string, mode ...DeleteMode) error {
+	err := w.filer.Delete(ctx, name, mode...)
+
+	// If the file is not found, it might be a notebook.
+	if errors.As(err, &FileDoesNotExistError{}) {
+		stat, ok := w.removeNotebookExtension(ctx, name)
+		if !ok {
+			// Not a valid notebook. Return the original error.
+			return err
+		}
+		return w.filer.Delete(ctx, stat.nameForWorkspaceAPI, mode...)
+	}
+
+	return err
+}
+
+// Try to stat the file as a regular file. If the file is not found, try to stat it as a notebook.
+func (w *workspaceFilesExtensionsClient) Stat(ctx context.Context, name string) (fs.FileInfo, error) {
+	info, err := w.filer.Stat(ctx, name)
+
+	// If the file is not found in the local file system, it might be a notebook.
```

Comment:
Why local file system? It seems like w.filer is WSFS filer if I read it correctly?

Reply:
It is, fixed the comment.